### PR TITLE
feat: udp connection

### PR DESCRIPTION
**Pull request by lchenut**
Wrap the DatagramTransport from chronos.
It's used to simply store the message and the remote address for the underlying protocols

### CODE REVIEW

File: webrtc/udp_connection.nim [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+# UdpConn is a small wrapper of the chronos DatagramTransport.
+# It's the simplest solution we found to store the message and
+# the remote address used by the underlying protocols (dtls/sctp etc...)
+
+type
+  UdpConn* = ref object
+    laddr*: TransportAddress
+    udp: DatagramTransport
+    dataRecv: AsyncQueue[(seq[byte], TransportAddress)]
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:23]: how about creating a type for this tuple?

**Comment by diegomrsantos** [comment]
it'd be great if you could add docs to the procs.

File: webrtc/udp_connection.nim [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    trace "UDP onReceive"
+    let msg = udp.getMessage()
+    self.dataRecv.addLastNoWait((msg, address))
+
+  self.dataRecv = newAsyncQueue[(seq[byte], TransportAddress)]()
+  self.udp = newDatagramTransport(onReceive, local = laddr)
+
+proc close*(self: UdpConn) {.async.} =
+  if self.closed:
+    debug "Try to close UdpConn twice"
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:40]: I think something like "Trying to close an already closed connection." would be better.

**Comment by diegomrsantos** [comment]
Could we use the support for exceptions in async procs?

File: webrtc/udp_connection.nim [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  ## Write a message on Udp to a remote address `raddr`
+  ##
+  if self.closed:
+    debug "Try to write on an already closed UdpConn"
+    return
+  trace "UDP write", msg
+  try:
+    await self.udp.sendTo(raddr, msg)
+  except TransportError as exc:
+    raise (ref WebRtcUdpError)(msg: exc.msg)
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:78]: I think we should raise a new `WebRtcUdpError` and use `TransportError` as the parent with `newException` proc.

Comment (lchenut) [reply webrtc/udp_connection.nim]: done :+1:

File: webrtc/udp_connection.nim [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  ## Write a message on Udp to a remote address `raddr`
+  ##
+  if self.closed:
+    debug "Try to write on an already closed UdpConn"
+    return
+  trace "UDP write", msg
+  try:
+    await self.udp.sendTo(raddr, msg)
+  except TransportError as exc:
+    raise newException(WebRtcUdpError, exc.msg, exc)
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:78]: I think the msg here should be related to the current error, something like "error when writing..."

File: webrtc/udp_connection.nim [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  except CancelledError as exc:
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:79]: If we catch and reraise without doing anything, I think we can just remove this.

File: webrtc/udp_connection.nim [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    raise exc
+
+proc read*(self: UdpConn): Future[UdpPacketInfo] {.async: (raises: [CancelledError]).} =
+  ## Read the next received Udp message
+  ##
+  if self.closed:
+    debug "Try to read on an already closed UdpConn"
+    return
+  trace "UDP read"
+  return await self.dataRecv.popFirst()
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:89]: No error can happen here?

File: webrtc/udp_connection.nim (hunk 5 above)
Comment (lchenut) [reply webrtc/udp_connection.nim:79]: Hum... You're right, it's a bad habit, catching/reraising is only useful when we catch a `CatchableError` before.

File: webrtc/udp_connection.nim (hunk 6 above)
Comment (lchenut) [reply webrtc/udp_connection.nim:89]: Only `CancelledError` : https://github.com/status-im/nim-chronos/blob/master/chronos/asyncsync.nim#L328-L329

File: webrtc/udp_connection.nim [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    message: seq[byte]
+    raddr: TransportAddress
+
+  UdpConn* = ref object
+    laddr*: TransportAddress
+    udp: DatagramTransport
+    dataRecv: AsyncQueue[UdpPacketInfo]
+    closed: bool
+
+proc init*(self: UdpConn, laddr: TransportAddress) =
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:33]: I think this is a constructor, isn't it? If so, it seems awkward to pass an instance to it. See more on https://status-im.github.io/nim-style-guide/language.objconstr.html

File: webrtc/udp_connection.nim (hunk 5 above)
Comment (lchenut) [reply webrtc/udp_connection.nim]: done

File: webrtc/udp_connection.nim (hunk 4 above)
Comment (lchenut) [reply webrtc/udp_connection.nim]: done

File: webrtc/udp_connection.nim [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    msg: seq[byte]
+  ) {.async: (raises: [CancelledError, WebRtcUdpError]).} =
+  ## Write a message on Udp to a remote address `raddr`
+  ##
+  if self.closed:
+    debug "Try to write on an already closed UdpConn"
+    return
+  trace "UDP write", msg
+  try:
+    await self.udp.sendTo(raddr, msg)
```
Comment (diegomrsantos) [inline webrtc/udp_connection.nim:76]: @arnetheduck In this case where we don't need to use the actual value that will be returned by the `Future`, should we do like it's done here or remove the await/async and return the Future from `sendTo`?

**Review by diegomrsantos — APPROVED** [review]

**Pull request merged**